File: examples/boolean_shapes.rb
require 'toxiclibs'
# Original toxi example had ellipse and circle but there were flaws in union and
# and XOR rendering caused by ellipse shapes
load_library :control_panel
attr_reader :gfx, :bool, :panel, :type, :hide, :polies
include Toxi

TYPE = [
  BooleanShapeBuilder::Type::UNION,
  BooleanShapeBuilder::Type::XOR
]
KEY = %w(union xor).freeze

def setup
  sketch_title 'Boolean Shapes'
  @gfx = Gfx::ToxiclibsSupport.new(self)
  @bool = KEY.zip(TYPE).to_h
  control_panel do |c|
    c.title = 'Union or XOR'
    c.menu :type, KEY, 'union'
    @panel = c
  end
  @hide = false
end

def draw
  unless hide
    @hide = true
    panel.set_visible(hide)
  end
  background(160)
  builder = BooleanShapeBuilder.new(bool[type])
  phi = frame_count * 0.01
  builder.add_shape(Circle.new(mouse_x, mouse_y, 70).toPolygon2D(30))
  builder.add_shape(Rect.new(200 + sin(phi * 13 / 8) * 50, 180, 100, 100))
  builder.add_shape(Triangle2D.create_equilateral_from(
    TVec2D.new(50 + sin(phi * 15 / 13) * 50, 200), TVec2D.new(300, 200))
  )
  builder.add_shape(
    Circle.new(100, 300, 50 + 30 * sin(phi * 21 / 15)).toPolygon2D(6)
  )
  fill(125)
  stroke(255, 0, 0)
  @polies = builder.compute_shapes
  polies.each { |p| gfx.polygon2D(p) }
end

def settings
  size(400, 400)
end

File: src/test/java/BooleanShapes.java

import processing.core.*;

import toxi.geom.*;
import toxi.processing.*;
import toxi.color.*;
import java.util.List;

public class BooleanShapes extends PApplet {

    ToxiclibsSupport gfx;

    @Override
    public void setup() {

        gfx = new ToxiclibsSupport(this);
    }

    @Override
    public void draw() {
        background(160);
        BooleanShapeBuilder builder = new BooleanShapeBuilder(
                BooleanShapeBuilder.Type.XOR);
        float phi = frameCount * 0.01f;
        builder.addShape(new Circle(mouseX, mouseY, 50));
        builder.addShape(new Ellipse(150, 130 + sin(phi) * 50, 120, 60));
        builder.addShape(new Rect(200 + sin(phi * 13 / 8) * 50, 180, 100, 100));
        builder.addShape(Triangle2D.createEquilateralFrom(new Vec2D(
                50 + sin(phi * 15 / 13) * 50, 200), new Vec2D(300, 200)));
        builder.addShape(new Circle(100, 300, 50 + 30 * sin(phi * 21 / 15))
                .toPolygon2D(6));
        noFill();
        stroke(255, 0, 0);
        List<Polygon2D> polies = builder.computeShapes();
        polies.forEach((p) -> {
            gfx.polygon2D(p);
        });
        noStroke();
        Vec2D p = new Vec2D();
        for (int y = 10; y < height; y += 20) {
            for (int x = 10; x < height; x += 20) {
                gfx.fill(pointInShapes(p.set(x, y), polies) ? TColor.GREEN
                        : TColor.BLUE);
                gfx.circle(p, 3);
            }
        }
    }

    public boolean pointInShapes(Vec2D p, List<Polygon2D> polies) {
        return polies.stream().anyMatch((poly) -> (poly.containsPoint(p)));
    }

    @Override
    public void settings() {
        size(400, 400);
    }

    static public void main(String[] passedArgs) {
        String[] appletArgs = new String[]{"BooleanShapes"};
        if (passedArgs != null) {
            PApplet.main(concat(appletArgs, passedArgs));
        } else {
            PApplet.main(appletArgs);
        }
    }
}
